File: agents/services/dynamodb-service.ts
/**
 * DynamoDB Service for Risk Scoring Agent
 *
 * Handles all DynamoDB operations for storing and retrieving risk scores,
 * reviewer feedback, and processing metadata.
 */

import {
  DynamoDBClient,
  PutItemCommand,
  GetItemCommand,
  QueryCommand,
  UpdateItemCommand,
  BatchWriteItemCommand,
} from '@aws-sdk/client-dynamodb';
import { marshall, unmarshall } from '@aws-sdk/util-dynamodb';
import {
  RiskScoreOutput,
  RiskScoreRecord,
  ReviewerFeedback,
} from '../types/risk-scoring.types';

export interface DynamoDBConfig {
  tableName: string;
  region: string;
}

export class DynamoDBService {
  private client: DynamoDBClient;
  private tableName: string;

  constructor(config: DynamoDBConfig) {
    this.client = new DynamoDBClient({ region: config.region });
    this.tableName = config.tableName;
  }

  /**
   * Save risk score to DynamoDB
   */
  async saveRiskScore(riskScore: RiskScoreOutput): Promise<void> {
    const timestamp = new Date().toISOString();

    const record: RiskScoreRecord = {
      pk: riskScore.contractId,
      sk: timestamp,
      contractId: riskScore.contractId,
      riskScore,
      createdAt: timestamp,
      updatedAt: timestamp,
      // Set TTL to 90 days from now (in seconds)
      ttl: Math.floor(Date.now() / 1000) + 90 * 24 * 60 * 60,
    };

    const command = new PutItemCommand({
      TableName: this.tableName,
      Item: marshall(record, {
        removeUndefinedValues: true,
        convertClassInstanceToMap: true,
      }),
    });

    try {
      await this.client.send(command);
      console.log(`Risk score saved for contract ${riskScore.contractId}`);
    } catch (error) {
      console.error('Error saving risk score to DynamoDB:', error);
      throw new Error(`Failed to save risk score: ${error.message}`);
    }
  }

  /**
   * Get risk score by contract ID and timestamp
   */
  async getRiskScore(contractId: string, timestamp?: string): Promise<RiskScoreRecord | null> {
    if (timestamp) {
      // Get specific version
      const command = new GetItemCommand({
        TableName: this.tableName,
        Key: marshall({
          pk: contractId,
          sk: timestamp,
        }),
      });

      try {
        const response = await this.client.send(command);
        if (!response.Item) {
          return null;
        }
        return unmarshall(response.Item) as RiskScoreRecord;
      } catch (error) {
        console.error('Error getting risk score from DynamoDB:', error);
        throw new Error(`Failed to get risk score: ${error.message}`);
      }
    } else {
      // Get latest version
      return this.getLatestRiskScore(contractId);
    }
  }

  /**
   * Get latest risk score for a contract
   */
  async getLatestRiskScore(contractId: string): Promise<RiskScoreRecord | null> {
    const command = new QueryCommand({
      TableName: this.tableName,
      KeyConditionExpression: 'pk = :contractId',
      ExpressionAttributeValues: marshall({
        ':contractId': contractId,
      }),
      ScanIndexForward: false, // Descending order (latest first)
      Limit: 1,
    });

    try {
      const response = await this.client.send(command);
      if (!response.Items || response.Items.length === 0) {
        return null;
      }
      return unmarshall(response.Items[0]) as RiskScoreRecord;
    } catch (error) {
      console.error('Error querying latest risk score:', error);
      throw new Error(`Failed to query risk score: ${error.message}`);
    }
  }

  /**
   * Get all risk scores for a contract (history)
   */
  async getRiskScoreHistory(contractId: string, limit: number = 10): Promise<RiskScoreRecord[]> {
    const command = new QueryCommand({
      TableName: this.tableName,
      KeyConditionExpression: 'pk = :contractId',
      ExpressionAttributeValues: marshall({
        ':contractId': contractId,
      }),
      ScanIndexForward: false, // Descending order (latest first)
      Limit: limit,
    });

    try {
      const response = await this.client.send(command);
      if (!response.Items || response.Items.length === 0) {
        return [];
      }
      return response.Items.map((item) => unmarshall(item) as RiskScoreRecord);
    } catch (error) {
      console.error('Error querying risk score history:', error);
      throw new Error(`Failed to query risk score history: ${error.message}`);
    }
  }

  /**
   * Add reviewer feedback to a risk score
   */
  async addReviewerFeedback(
    contractId: string,
    timestamp: string,
    feedback: ReviewerFeedback
  ): Promise<void> {
    const command = new UpdateItemCommand({
      TableName: this.tableName,
      Key: marshall({
        pk: contractId,
        sk: timestamp,
      }),
      UpdateExpression: 'SET reviewerFeedback = :feedback, updatedAt = :updatedAt',
      ExpressionAttributeValues: marshall({
        ':feedback': feedback,
        ':updatedAt': new Date().toISOString(),
      }),
    });

    try {
      await this.client.send(command);
      console.log(`Reviewer feedback added for contract ${contractId}`);
    } catch (error) {
      console.error('Error adding reviewer feedback:', error);
      throw new Error(`Failed to add reviewer feedback: ${error.message}`);
    }
  }

  /**
   * Update human loop status
   */
  async updateHumanLoopStatus(
    contractId: string,
    humanLoopArn: string,
    status: string
  ): Promise<void> {
    // Get latest record first
    const latestRecord = await this.getLatestRiskScore(contractId);
    if (!latestRecord) {
      throw new Error(`No risk score found for contract ${contractId}`);
    }

    const command = new UpdateItemCommand({
      TableName: this.tableName,
      Key: marshall({
        pk: contractId,
        sk: latestRecord.sk,
      }),
      UpdateExpression: 'SET humanLoopArn = :arn, humanLoopStatus = :status, updatedAt = :updatedAt',
      ExpressionAttributeValues: marshall({
        ':arn': humanLoopArn,
        ':status': status,
        ':updatedAt': new Date().toISOString(),
      }),
    });

    try {
      await this.client.send(command);
      console.log(`Human loop status updated for contract ${contractId}: ${status}`);
    } catch (error) {
      console.error('Error updating human loop status:', error);
      throw new Error(`Failed to update human loop status: ${error.message}`);
    }
  }

  /**
   * Batch save multiple risk scores
   */
  async batchSaveRiskScores(riskScores: RiskScoreOutput[]): Promise<void> {
    const timestamp = new Date().toISOString();
    const ttl = Math.floor(Date.now() / 1000) + 90 * 24 * 60 * 60;

    // DynamoDB batch write supports max 25 items
    const batches: RiskScoreRecord[][] = [];
    for (let i = 0; i < riskScores.length; i += 25) {
      batches.push(
        riskScores.slice(i, i + 25).map((riskScore) => ({
          pk: riskScore.contractId,
          sk: timestamp,
          contractId: riskScore.contractId,
          riskScore,
          createdAt: timestamp,
          updatedAt: timestamp,
          ttl,
        }))
      );
    }

    for (const batch of batches) {
      const command = new BatchWriteItemCommand({
        RequestItems: {
          [this.tableName]: batch.map((record) => ({
            PutRequest: {
              Item: marshall(record, {
                removeUndefinedValues: true,
                convertClassInstanceToMap: true,
              }),
            },
          })),
        },
      });

      try {
        await this.client.send(command);
        console.log(`Batch saved ${batch.length} risk scores`);
      } catch (error) {
        console.error('Error batch saving risk scores:', error);
        throw new Error(`Failed to batch save risk scores: ${error.message}`);
      }
    }
  }

  /**
   * Query high-risk contracts
   */
  async queryHighRiskContracts(
    minScore: number = 0.7,
    limit: number = 100
  ): Promise<RiskScoreRecord[]> {
    // Note: This requires a GSI on compositeScore
    const command = new QueryCommand({
      TableName: this.tableName,
      IndexName: 'CompositeScoreIndex', // Assumes GSI exists
      KeyConditionExpression: 'scoreType = :scoreType',
      FilterExpression: 'riskScore.compositeScore >= :minScore',
      ExpressionAttributeValues: marshall({
        ':scoreType': 'COMPOSITE',
        ':minScore': minScore,
      }),
      ScanIndexForward: false,
      Limit: limit,
    });

    try {
      const response = await this.client.send(command);
      if (!response.Items || response.Items.length === 0) {
        return [];
      }
      return response.Items.map((item) => unmarshall(item) as RiskScoreRecord);
    } catch (error) {
      console.error('Error querying high-risk contracts:', error);
      // If GSI doesn't exist, return empty array
      console.warn('CompositeScoreIndex may not exist. Consider creating a GSI for efficient queries.');
      return [];
    }
  }

  /**
   * Get statistics for risk scores
   */
  async getRiskScoreStats(startDate: string, endDate: string): Promise<any> {
    // This would typically use DynamoDB Streams or be computed via analytics
    // For now, return a placeholder
    console.log('Getting risk score statistics:', { startDate, endDate });
    return {
      totalContracts: 0,
      avgCompositeScore: 0,
      highRiskCount: 0,
      mediumRiskCount: 0,
      lowRiskCount: 0,
    };
  }
}

File: agents-python/services/dynamodb_service.py
"""
DynamoDB Service for Risk Scoring Agent

Handles all DynamoDB operations for storing and retrieving risk scores,
reviewer feedback, and processing metadata.
"""

import boto3
from boto3.dynamodb.conditions import Key
from datetime import datetime, timedelta
from typing import List, Optional
import json

from types.risk_scoring_types import (
    RiskScoreOutput,
    RiskScoreRecord,
    ReviewerFeedback
)


class DynamoDBService:
    """DynamoDB service for risk score storage"""

    def __init__(self, table_name: str, region: str = "us-east-1"):
        """
        Initialize DynamoDB service

        Args:
            table_name: DynamoDB table name
            region: AWS region
        """
        self.table_name = table_name
        self.region = region
        self.dynamodb = boto3.resource('dynamodb', region_name=region)
        self.table = self.dynamodb.Table(table_name)

    def save_risk_score(self, risk_score: RiskScoreOutput) -> None:
        """
        Save risk score to DynamoDB

        Args:
            risk_score: Risk score output to save
        """
        timestamp = datetime.utcnow().isoformat()
        ttl = int((datetime.utcnow() + timedelta(days=90)).timestamp())

        record = RiskScoreRecord(
            pk=risk_score.contract_id,
            sk=timestamp,
            contract_id=risk_score.contract_id,
            risk_score=risk_score,
            created_at=timestamp,
            updated_at=timestamp,
            ttl=ttl
        )

        try:
            self.table.put_item(Item=json.loads(record.json()))
            print(f"Risk score saved for contract {risk_score.contract_id}")
        except Exception as e:
            print(f"Error saving risk score to DynamoDB: {e}")
            raise Exception(f"Failed to save risk score: {str(e)}")

    def get_risk_score(
        self,
        contract_id: str,
        timestamp: Optional[str] = None
    ) -> Optional[RiskScoreRecord]:
        """
        Get risk score by contract ID and timestamp

        Args:
            contract_id: Contract identifier
            timestamp: Specific timestamp (optional, gets latest if not provided)

        Returns:
            RiskScoreRecord or None if not found
        """
        if timestamp:
            # Get specific version
            try:
                response = self.table.get_item(
                    Key={'pk': contract_id, 'sk': timestamp}
                )
                if 'Item' in response:
                    return RiskScoreRecord(**response['Item'])
                return None
            except Exception as e:
                print(f"Error getting risk score from DynamoDB: {e}")
                raise Exception(f"Failed to get risk score: {str(e)}")
        else:
            # Get latest version
            return self.get_latest_risk_score(contract_id)

    def get_latest_risk_score(self, contract_id: str) -> Optional[RiskScoreRecord]:
        """
        Get latest risk score for a contract

        Args:
            contract_id: Contract identifier

        Returns:
            Latest RiskScoreRecord or None
        """
        try:
            response = self.table.query(
                KeyConditionExpression=Key('pk').eq(contract_id),
                ScanIndexForward=False,  # Descending order (latest first)
                Limit=1
            )

            if response['Items']:
                return RiskScoreRecord(**response['Items'][0])
            return None
        except Exception as e:
            print(f"Error querying latest risk score: {e}")
            raise Exception(f"Failed to query risk score: {str(e)}")

    def get_risk_score_history(
        self,
        contract_id: str,
        limit: int = 10
    ) -> List[RiskScoreRecord]:
        """
        Get all risk scores for a contract (history)

        Args:
            contract_id: Contract identifier
            limit: Maximum number of records to return

        Returns:
            List of RiskScoreRecord
        """
        try:
            response = self.table.query(
                KeyConditionExpression=Key('pk').eq(contract_id),
                ScanIndexForward=False,  # Descending order (latest first)
                Limit=limit
            )

            return [RiskScoreRecord(**item) for item in response.get('Items', [])]
        except Exception as e:
            print(f"Error querying risk score history: {e}")
            raise Exception(f"Failed to query risk score history: {str(e)}")

    def add_reviewer_feedback(
        self,
        contract_id: str,
        timestamp: str,
        feedback: ReviewerFeedback
    ) -> None:
        """
        Add reviewer feedback to a risk score

        Args:
            contract_id: Contract identifier
            timestamp: Risk score timestamp
            feedback: Reviewer feedback
        """
        try:
            self.table.update_item(
                Key={'pk': contract_id, 'sk': timestamp},
                UpdateExpression='SET reviewer_feedback = :feedback, updated_at = :updated',
                ExpressionAttributeValues={
                    ':feedback': json.loads(feedback.json()),
                    ':updated': datetime.utcnow().isoformat()
                }
            )
            print(f"Reviewer feedback added for contract {contract_id}")
        except Exception as e:
            print(f"Error adding reviewer feedback: {e}")
            raise Exception(f"Failed to add reviewer feedback: {str(e)}")

    def update_human_loop_status(
        self,
        contract_id: str,
        human_loop_arn: str,
        status: str
    ) -> None:
        """
        Update human loop status

        Args:
            contract_id: Contract identifier
            human_loop_arn: Amazon A2I human loop ARN
            status: Human loop status
        """
        # Get latest record first
        latest_record = self.get_latest_risk_score(contract_id)
        if not latest_record:
            raise Exception(f"No risk score found for contract {contract_id}")

        try:
            self.table.update_item(
                Key={'pk': contract_id, 'sk': latest_record.sk},
                UpdateExpression='SET human_loop_arn = :arn, human_loop_status = :status, updated_at = :updated',
                ExpressionAttributeValues={
                    ':arn': human_loop_arn,
                    ':status': status,
                    ':updated': datetime.utcnow().isoformat()
                }
            )
            print(f"Human loop status updated for contract {contract_id}: {status}")
        except Exception as e:
            print(f"Error updating human loop status: {e}")
            raise Exception(f"Failed to update human loop status: {str(e)}")

    def batch_save_risk_scores(self, risk_scores: List[RiskScoreOutput]) -> None:
        """
        Batch save multiple risk scores

        Args:
            risk_scores: List of risk scores to save
        """
        timestamp = datetime.utcnow().isoformat()
        ttl = int((datetime.utcnow() + timedelta(days=90)).timestamp())

        # DynamoDB batch write supports max 25 items
        for i in range(0, len(risk_scores), 25):
            batch = risk_scores[i:i + 25]

            with self.table.batch_writer() as writer:
                for risk_score in batch:
                    record = RiskScoreRecord(
                        pk=risk_score.contract_id,
                        sk=timestamp,
                        contract_id=risk_score.contract_id,
                        risk_score=risk_score,
                        created_at=timestamp,
                        updated_at=timestamp,
                        ttl=ttl
                    )
                    writer.put_item(Item=json.loads(record.json()))

            print(f"Batch saved {len(batch)} risk scores")

    def query_high_risk_contracts(
        self,
        min_score: float = 0.7,
        limit: int = 100
    ) -> List[RiskScoreRecord]:
        """
        Query high-risk contracts

        Args:
            min_score: Minimum composite score
            limit: Maximum number of records

        Returns:
            List of high-risk contract records
        """
        # Note: This requires a GSI on composite_score
        # For now, scan and filter (not efficient for large datasets)
        try:
            response = self.table.scan(
                FilterExpression='risk_score.composite_score >= :min_score',
                ExpressionAttributeValues={':min_score': min_score},
                Limit=limit
            )

            return [RiskScoreRecord(**item) for item in response.get('Items', [])]
        except Exception as e:
            print(f"Error querying high-risk contracts: {e}")
            print("Consider creating a GSI on composite_score for efficient queries")
            return []

    def get_risk_score_stats(self, start_date: str, end_date: str) -> dict:
        """
        Get statistics for risk scores

        Args:
            start_date: Start date (ISO format)
            end_date: End date (ISO format)

        Returns:
            Dictionary with statistics
        """
        # This would typically use DynamoDB Streams or be computed via analytics
        # Placeholder implementation
        print(f"Getting risk score statistics: {start_date} to {end_date}")
        return {
            'total_contracts': 0,
            'avg_composite_score': 0.0,
            'high_risk_count': 0,
            'medium_risk_count': 0,
            'low_risk_count': 0
        }
